### PR TITLE
[FEATURE] - added firebase auth and a simple ui for signin page

### PR DESCRIPTION
**Pull request by Anuraj-TR**
notes - need to fix default page,page layout,code structure and folder structure

SIGNIN PAGE -

image

**************************

dashboard page 

image

### CODE REVIEW

**Comment by anandureghu** [comment]
in the Layout component, **show only children if the location path is in '/signin'** otherwise show the current implementation

**Comment by Anuraj-TR** [comment]
signIn page 
 
image

******************************

dashboard page

image

**Comment by anandureghu** [comment]
change login butotn style

**Comment by anandureghu** [comment]
do we need login button on dashboard?

**Comment by Anuraj-TR** [comment]
Removed login button from dashboard added new styles to logout button

**desktop view** 
********
**default**

image

**active**

image

**************

**Mobile view**

image

**Comment by Anuraj-TR** [comment]
implemented user first name on navbar & added waving icon

image

******
**changes** 

used useEffect for setting userName and used react-icons

File: src/main/container/Navbar.jsx [hunk 1]
```diff
@@ -1,3 +1,52 @@
...
+import { useEffect, useState } from "react";
+
 export const Navbar = () => {
-  return <div>Header</div>;
+  const { user, logOut } = UserAuth();
+  const [userName, setUserName] = useState('')
+
+  const handleLogOut = async () => {
+    try {
+      await logOut()
```
Comment (anandureghu) [inline src/main/container/Navbar.jsx]: not an async function, dont use await

File: src/main/container/Navbar.jsx [hunk 2]
```diff
@@ -1,3 +1,52 @@
...
+import { RiLogoutCircleRLine } from "react-icons/ri";
+import { MdWavingHand } from "react-icons/md";
+import { useEffect, useState } from "react";
+
 export const Navbar = () => {
-  return <div>Header</div>;
+  const { user, logOut } = UserAuth();
+  const [userName, setUserName] = useState('')
+
+  const handleLogOut = async () => {
```
Comment (anandureghu) [inline src/main/container/Navbar.jsx]: dont need seperate handleLogout, just call logOut from context

File: src/main/container/Navbar.jsx [hunk 3]
```diff
@@ -1,3 +1,52 @@
...
+    try {
+      await logOut()
+    } catch (error) {
+      console.log(error);
+    }
+  }
+
+  useEffect(() => {
+    setUserName(user?.displayName.split(' ')[0])
```
Comment (anandureghu) [inline src/main/container/Navbar.jsx]: try seperating userDetails from authContext after login

File: src/main/context/AuthContext.jsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+  // signout
+  const logOut = () => {
+    signOut(auth)
+  }
+
+  // onAuthChange
+  useEffect(() => {
+    const unSubscribe = onAuthStateChanged(auth, (currentUser) => {
+      setUser(currentUser)
```
Comment (anandureghu) [inline src/main/context/AuthContext.jsx]: after getting currentUser, collect necessary data from currentUser and only set those details in the user state and store the full currentUser Details in a seperate useState like firebaseUser

_necessary data eg: firstName, lastName, email, imageUrl ...etc_

File: src/main/pages/SignUpPage.jsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import { UserAuth } from "../context/AuthContext";
+import { useNavigate } from "react-router-dom";
+
+export const SignUpPage = () => {
+  const { googleSignIn, user } = UserAuth();
+  const navigate = useNavigate()
+
+  const handleGoogleSignIn = async () => {
+    try {
+      await googleSignIn()
```
Comment (anandureghu) [inline src/main/pages/SignUpPage.jsx]: not an async function

File: src/main/pages/SignUpPage.jsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import { useEffect } from "react";
+import { UserAuth } from "../context/AuthContext";
+import { useNavigate } from "react-router-dom";
+
+export const SignUpPage = () => {
+  const { googleSignIn, user } = UserAuth();
+  const navigate = useNavigate()
+
+  const handleGoogleSignIn = async () => {
```
Comment (anandureghu) [inline src/main/pages/SignUpPage.jsx]: dont need seperate handleSignin Function

File: src/main/context/AuthContext.jsx [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+
+  return (
+    <AuthContext.Provider value={{googleSignIn,user,logOut}}>
+      {children}
+    </AuthContext.Provider>
+  )
+}
+
+export const UserAuth = () => {
```
Comment (anandureghu) [inline src/main/context/AuthContext.jsx]: move this into src>hooks>useAuth.js

_this is called custom hooks_

**Review by anandureghu — CHANGES_REQUESTED** [review]

File: src/main/container/Navbar.jsx (hunk 1 above)
Comment (Anuraj-TR) [reply src/main/container/Navbar.jsx]: removed handleLogout function

File: src/main/container/Navbar.jsx (hunk 2 above)
Comment (Anuraj-TR) [reply src/main/container/Navbar.jsx]: removed handleLogout function

File: src/main/pages/SignUpPage.jsx (hunk 5 above)
Comment (Anuraj-TR) [reply src/main/pages/SignUpPage.jsx]: removed handleGoogleSignIN function

File: src/main/pages/SignUpPage.jsx (hunk 6 above)
Comment (Anuraj-TR) [reply src/main/pages/SignUpPage.jsx]: removed handleGoogleSignIN function

File: src/main/context/AuthContext.jsx (hunk 7 above)
Comment (Anuraj-TR) [reply src/main/context/AuthContext.jsx]: moved to hooks

File: src/main/container/Navbar.jsx (hunk 3 above)
Comment (Anuraj-TR) [reply src/main/container/Navbar.jsx]: separated userDetails

File: src/main/context/AuthContext.jsx (hunk 4 above)
Comment (Anuraj-TR) [reply src/main/context/AuthContext.jsx]: created **firebaseUser** and added full **currentUser** details to it

changed user data to new obj 

image

File: src/main/context/AuthContext.jsx [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  // onAuthChange
+  useEffect(() => {
+    const unSubscribe = onAuthStateChanged(auth, (currentUser) => {
+      setFirebaseUser(currentUser);
+      setUser({
+        userId: currentUser.uid,
+        firstName: currentUser.displayName.split(" ")[0],
+        secondName: currentUser.displayName.split(" ")[1],
+        email: currentUser.email,
+        url: currentUser.photoURL,
```
Comment (anandureghu) [inline src/main/context/AuthContext.jsx]: change key name: imageUrl or photoUrl

File: src/main/context/AuthContext.jsx [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        phNumber: currentUser.phoneNumber,
```
Comment (anandureghu) [inline src/main/context/AuthContext.jsx]: phoneNumber is better i think, optional change

**Review by anandureghu — CHANGES_REQUESTED** [review]

File: src/main/context/AuthContext.jsx [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+  // google signin
+  const googleSignIn = () => {
+    const provider = new GoogleAuthProvider();
+    signInWithPopup(auth, provider)
+  }
+
+  // signout
+  const logOut = () => {
+    signOut(auth)
```
Comment (anandureghu) [inline src/main/context/AuthContext.jsx:23]: clear currentUser and user values from state after login and navigate to login page

**Review by anandureghu — CHANGES_REQUESTED** [review]

File: src/main/context/AuthContext.jsx (hunk 8 above)
Comment (Anuraj-TR) [reply src/main/context/AuthContext.jsx]: changed to imageUrl

File: src/main/context/AuthContext.jsx (hunk 10 above)
Comment (Anuraj-TR) [reply src/main/context/AuthContext.jsx:23]: cleared user values and added navigate to signin 

need help with restricting user from entering the dashboard page without login.

File: src/main/context/AuthContext.jsx (hunk 9 above)
Comment (Anuraj-TR) [reply src/main/context/AuthContext.jsx]: changed to phoneNumber

File: src/main/context/AuthContext.jsx (hunk 10 above)
Comment (anandureghu) [reply src/main/context/AuthContext.jsx:23]: refer Protected Routes

File: src/main/context/AuthContext.jsx [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    setUser(null)
```
Comment (anandureghu) [inline src/main/context/AuthContext.jsx:24]: setFirebaseUser also null

**Review by anandureghu — CHANGES_REQUESTED** [review]

File: src/main/context/AuthContext.jsx (hunk 11 above)
Comment (Anuraj-TR) [reply src/main/context/AuthContext.jsx:24]: done

**Review by anandureghu — APPROVED** [review]

**Pull request merged**